presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -508,21 +508,45 @@ public PlanWithProperties visitTableWriter(TableWriterNode originalTableWriterNo
             PlanWithProperties tableWriter;
 
             if (!originalTableWriterNode.getTablePartitioningScheme().isPresent()) {
-                tableWriter = planAndEnforceChildren(
-                        new TableWriterNode(
-                                originalTableWriterNode.getId(),
-                                originalTableWriterNode.getSource(),
-                                originalTableWriterNode.getTarget(),
-                                variableAllocator.newVariable("partialrowcount", BIGINT),
-                                variableAllocator.newVariable("partialfragments", VARBINARY),
-                                variableAllocator.newVariable("partialcontext", VARBINARY),
-                                originalTableWriterNode.getColumns(),
-                                originalTableWriterNode.getColumnNames(),
-                                originalTableWriterNode.getTablePartitioningScheme(),
-                                originalTableWriterNode.getPreferredShufflePartitioningScheme(),
-                                statisticAggregations.map(StatisticAggregations.Parts::getPartialAggregation)),
-                        fixedParallelism(),
-                        fixedParallelism());
+                int taskWriterCount = getTaskWriterCount(session);
+                int taskConcurrency = getTaskConcurrency(session);
+                if (taskWriterCount == taskConcurrency) {
+                    tableWriter = planAndEnforceChildren(
+                            new TableWriterNode(
+                                    originalTableWriterNode.getId(),
+                                    originalTableWriterNode.getSource(),
+                                    originalTableWriterNode.getTarget(),
+                                    variableAllocator.newVariable("partialrowcount", BIGINT),
+                                    variableAllocator.newVariable("partialfragments", VARBINARY),
+                                    variableAllocator.newVariable("partialcontext", VARBINARY),
+                                    originalTableWriterNode.getColumns(),
+                                    originalTableWriterNode.getColumnNames(),
+                                    originalTableWriterNode.getTablePartitioningScheme(),
+                                    originalTableWriterNode.getPreferredShufflePartitioningScheme(),
+                                    statisticAggregations.map(StatisticAggregations.Parts::getPartialAggregation)),
+                            fixedParallelism(),
```

Comment:
Could you explain why it is fixedParallelism() currently? I am trying to understand the difference for ` public enum StreamDistribution
        {
            SINGLE, MULTIPLE, FIXED
        }`

Reply:
- `SINGLE` - single operator instance is required (e.g.: OrderByOperator, TopNOperator, TableCommitOperator, etc.)
- `FIXED` - fixed number of operator instances is required (e.g.: HashAggregationOperator, WindowOperator, etc)
- `MULTIPLE` - there's no requirement for the number of operator instances (e.g.: ScanFiterProject Operator).


> Could you explain why it is fixedParallelism() currently?

We request fixed parallelism to ensure that the number of TableWriterOperators is fixed for a given task, so it produces a fixed number of output files.

Reply:
Thanks for explanation.

Reply:
@arhimondr : 

> MULTIPLE - there's no requirement for the number of operator instances (e.g.: ScanFiterProject Operator).

So essentially there will be one driver per each split?

Reply:
> FIXED - fixed number of operator instances is required (e.g.: HashAggregationOperator, WindowOperator, etc)

What will be the parallelism for FIXED? 
https://github.com/prestodb/presto/blob/6970b6da9f392a8f6eb41c974c7eeb15f69c62da/presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPreferredProperties.java#L78-L81

Is it `taskConcurrency` ?

Reply:
> So essentially there will be one driver per each split?

Yeah. And the number of divers run in parallel for scan is controlled by a different set of properties (e.g.: `task.max-drivers-per-task`)